### PR TITLE
Fix StatefulSet MaxUnavailable with MinReadySeconds, and add necessary e2e tests

### LINKED ISSUE #112307: Parallel pod management with maxUnavailable set should not ignore minReadySeconds
### What happened?

StatefulSet spec's template was updated:

<details>

```apiVersion: v1
items:
- apiVersion: apps/v1
  kind: StatefulSet
  # many other unrelated stuff
  spec:
    replicas: 5
    minReadySeconds: 300
    podManagementPolicy: Parallel
    updateStrategy:
      rollingUpdate:
        maxUnavailable: 1
```
</details>

Kubernetes has updated all pods one-by-one in a quick succession, the next pod was updated as soon as previous one has passed the startup probe.
`status.availableReplicas` has dropped from 5 to 0 over the course of the push (which took ≈3 minutes).

### What did you expect to happen?

I did expected that kubernetes would update pods one-by-one, taking `minReadySeconds` into account, the next pod was expected to be updated `minReadySeconds` seconds after previous one has passed the startup probe.
`status.availableReplicas` should not have been dropped below 4 at any point (as `maxUnavailable` is explicitly set to 1).

### How can we reproduce it (as minimally and precisely as possible)?

Create a `StatefulSet` with the spec above, and update its `spec.template` field to trigger a rollout.
`MaxUnavailableStatefulSet` and `StatefulSetMinReadySeconds` feature gates must be enabled.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.4-dbx2", GitCommit:"95ee5ab382d64cfe6c28967f36b53970b8374491", GitTreeState:"clean", BuildDate:"2001-09-09T01:46:40Z", GoVersion:"go1.18.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.4-dbx2", GitCommit:"95ee5ab382d64cfe6c28967f36b53970b8374491", GitTreeState:"clean", BuildDate:"2001-09-09T01:46:40Z", GoVersion:"go1.18.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
In-house kubernetes setup.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux $hostname 5.15.0-39-generic #42-Ubuntu SMP Thu Jun 9 23:42:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by leomichalski**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This is necessary to promote this feature from alpha to beta. StatefulSet MaxUnavailable needs to honor MinReadySeconds, not ignore it. 

#### Which issue(s) this PR fixes:
Fixes #112307

#### Special notes for your reviewer:
Relates to https://github.com/kubernetes/enhancements/pull/4474

Addresses https://github.com/kubernetes/enhancements/pull/4474#discussion_r1478474701

#### Does this PR introduce a user-facing change?

```release-note
StatefulSet MaxUnavailable now properly honors MinReadySeconds.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/pull/4474

### CODE REVIEW

**Comment by leomichalski** [comment]
I changed this line temporarily so only this test is ran.

From this
https://github.com/leomichalski/kubernetes/blob/bfce32798dbaef181c3952a919026d15ace0d112/test/e2e/apps/statefulset.go#L1674

To that
```
var _ = SIGDescribe("Rolling update strategy with MaxUnavailable", ginkgo.Focus, func() {
```

Test command:

```bash
go test -v ./test/e2e -ginkgo.v -kubeconfig=/home/lab/.kube/config
```

Test logs without the fix (1 test fails and 2 pass, as they should)

```
  I0313 11:05:41.146320 119445 test_context.go:561] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
=== RUN   TestE2E
  I0313 11:05:41.146442  119445 e2e.go:109] Starting e2e run "9655e93e-fd18-440b-afad-54c7b0368250" on Ginkgo node 1
Running Suite: Kubernetes e2e suite - /home/lab/Desktop/pareamento/kubernetes/test/e2e
======================================================================================
Random Seed: 1710338740 - will randomize all specs

Will run 3 of 7197 specs
------------------------------
[ReportBeforeSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e_test.go:157
[ReportBeforeSuite] PASSED [0.000 seconds]
------------------------------
[SynchronizedBeforeSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e.go:69
  I0313 11:05:41.286008 119445 util.go:506] >>> kubeConfig: /home/lab/.kube/config
  I0313 11:05:41.287699 119445 helper.go:48] Waiting up to 30m0s for all (but 0) nodes to be schedulable
  I0313 11:05:41.304627 119445 e2e.go:142] Waiting up to 5m0s for all daemonsets in namespace 'kube-system' to start
  I0313 11:05:41.306530 119445 e2e.go:153] 1 / 1 pods ready in namespace 'kube-system' in daemonset 'kindnet' (0 seconds elapsed)
  I0313 11:05:41.306559 119445 e2e.go:153] 1 / 1 pods ready in namespace 'kube-system' in daemonset 'kube-proxy' (0 seconds elapsed)
  I0313 11:05:41.306568 119445 e2e.go:245] e2e test version: v0.0.0-master+$Format:%H$
  I0313 11:05:41.307072 119445 e2e.go:254] kube-apiserver version: v1.27.3
  I0313 11:05:41.307120 119445 util.go:506] >>> kubeConfig: /home/lab/.kube/config
  I0313 11:05:41.309017 119445 e2e.go:383] Cluster IP family: ipv4
[SynchronizedBeforeSuite] PASSED [0.023 seconds]
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should update successfully with OrderedReady PodManagementPolicy [sig-apps]

(...)

------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should honor MinReadySeconds [sig-apps]
(...)

------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should update successfully with Parallel PodManagementPolicy [sig-apps]
(...)

------------------------------
[SynchronizedAfterSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e.go:80
  I0313 11:09:02.622261 119445 suites.go:34] Running AfterSuite actions on node 1
[SynchronizedAfterSuite] PASSED [0.000 seconds]
------------------------------
[ReportAfterSuite] Kubernetes e2e suite report
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e_test.go:161
[ReportAfterSuite] PASSED [0.000 seconds]
------------------------------

Summarizing 1 Failure:
  [FAIL] [sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature [It] should honor MinReadySeconds [sig-apps]
  /home/lab/Desktop/pareamento/kubernetes/test/e2e/apps/statefulset.go:1737

Ran 3 of 7197 Specs in 201.336 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 7194 Skipped
--- FAIL: TestE2E (201.48s)
FAIL
FAIL	k8s.io/kubernetes/test/e2e	201.889s
FAIL
```

---
---
---

Test logs with the fix (all tests pass)

```
  I0313 11:37:50.426229 158105 test_context.go:561] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
=== RUN   TestE2E
  I0313 11:37:50.426346  158105 e2e.go:109] Starting e2e run "9e852e18-c8e8-46e7-a983-d4eecc53ef50" on Ginkgo node 1
Running Suite: Kubernetes e2e suite - /home/lab/Desktop/pareamento/kubernetes/test/e2e
======================================================================================
Random Seed: 1710340670 - will randomize all specs

Will run 3 of 7197 specs
------------------------------
[ReportBeforeSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e_test.go:157
[ReportBeforeSuite] PASSED [0.000 seconds]
------------------------------
[SynchronizedBeforeSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e.go:69
  I0313 11:37:50.571733 158105 util.go:506] >>> kubeConfig: /home/lab/.kube/config
  I0313 11:37:50.572734 158105 helper.go:48] Waiting up to 30m0s for all (but 0) nodes to be schedulable
  I0313 11:37:50.586563 158105 e2e.go:142] Waiting up to 5m0s for all daemonsets in namespace 'kube-system' to start
  I0313 11:37:50.589395 158105 e2e.go:153] 1 / 1 pods ready in namespace 'kube-system' in daemonset 'kindnet' (0 seconds elapsed)
  I0313 11:37:50.589449 158105 e2e.go:153] 1 / 1 pods ready in namespace 'kube-system' in daemonset 'kube-proxy' (0 seconds elapsed)
  I0313 11:37:50.589464 158105 e2e.go:245] e2e test version: v0.0.0-master+$Format:%H$
  I0313 11:37:50.590674 158105 e2e.go:254] kube-apiserver version: v1.27.3
  I0313 11:37:50.590785 158105 util.go:506] >>> kubeConfig: /home/lab/.kube/config
  I0313 11:37:50.593755 158105 e2e.go:383] Cluster IP family: ipv4
[SynchronizedBeforeSuite] PASSED [0.022 seconds]
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should honor MinReadySeconds [sig-apps]
(...)

------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should update successfully with Parallel PodManagementPolicy [sig-apps]
(...)

------------------------------
[sig-apps] Rolling update strategy with MaxUnavailable Rolling update with the MaxUnavailable feature should update successfully with OrderedReady PodManagementPolicy [sig-apps]
(...)

------------------------------
[SynchronizedAfterSuite] 
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e.go:80
  I0313 11:42:31.762832 158105 suites.go:34] Running AfterSuite actions on node 1
[SynchronizedAfterSuite] PASSED [0.000 seconds]
------------------------------
[ReportAfterSuite] Kubernetes e2e suite report
/home/lab/Desktop/pareamento/kubernetes/test/e2e/e2e_test.go:161
[ReportAfterSuite] PASSED [0.000 seconds]
------------------------------

Ran 3 of 7197 Specs in 281.191 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 7194 Skipped
PASS | FOCUSED
FAIL	k8s.io/kubernetes/test/e2e	281.786s
FAIL

```

**Comment by leomichalski** [comment]
Instructions to reproduce the tests


#### Setup

Setup a cluster with "MaxUnavailableStatefulSet" feature gate enabled. The following Kind config may be used.

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  MaxUnavailableStatefulSet: true
```

Change this line (https://github.com/leomichalski/kubernetes/blob/bfce32798dbaef181c3952a919026d15ace0d112/test/e2e/apps/statefulset.go#L1674) temporarily so only the right e2e test runs.

```go
// from
var _ = SIGDescribe("Rolling update strategy with MaxUnavailable", feature.MaxUnavailableStatefulSet, func() {


// to
var _ = SIGDescribe("Rolling update strategy with MaxUnavailable", ginkgo.Focus, func() {
```

#### Run the test without the fix

Run the test

```bash
go test -v ./test/e2e -ginkgo.v -kubeconfig=/home/lab/.kube/config
```

#### Run the test with the fix

Do the fix.

```bash
# Clone this branch.
git clone https://github.com/leomichalski/kubernetes -b fix/minreadyseconds-maxunavailable

# Build the kube controller manager.
make quick-release-images DBG=1

# Load the new image to the Docker registry
docker load -i _output/release-images/amd64/kube-controller-manager.tar

# Load Docker image to Kind substituting "SUBSTITUTE_WITH_TAG"
kind load docker-image registry.k8s.io/kube-controller-manager-amd64:SUBSTITUTE_WITH_TAG

# Edit `/etc/kubernetes/manifests/kube-controller-manager.yaml` with your favorite editor so it uses the new Docker image

# If necessary, delete the old kube controller manager pod so it starts a new one using the new Docker image
```

Run the test

```bash
go test -v ./test/e2e -ginkgo.v -kubeconfig=/home/lab/.kube/config
```

File: pkg/controller/statefulset/stateful_set_control.go [hunk 1]
```diff
@@ -759,7 +759,7 @@ func updateStatefulSetAfterInvariantEstablished(
 	// those pods and count the successful deletions. Update the status with the correct number of deletions.
 	unavailablePods := 0
 	for target := len(replicas) - 1; target >= 0; target-- {
-		if !isHealthy(replicas[target]) {
+		if !isRunningAndAvailable(replicas[target], set.Spec.MinReadySeconds) || isTerminating(replicas[target]) {
```
Comment (tahirali-csc) [inline pkg/controller/statefulset/stateful_set_control.go:762]: Why we need to consider `isTerminating`?

Comment (leomichalski) [reply pkg/controller/statefulset/stateful_set_control.go:762]: Terminating is not a phase (such as Pending, Running, Succeeded, Failed, Unknown). Under the hoods, `isTerminating` just checks if the pod DeletionTimestamp is not nil. Current `isHealthy` implementation (`isRunningAndReady(pod) && !isTerminating(pod)`) is quite similar to the line that substituted `isHealthy`. But, there isn't a method (like `isHealthy`) that encapsulates that logic for `isRunningAndAvailable` yet.

There are some **very very** related bugs (on which I'm working on for the next few days) in "StatefulSetMinReadySeconds" feature. They're related to `isRunningAndAvailable` and `isRunningAndReady` methods. I'll probably focus on them, unblocking KEP-961 along the way. What the `isRunningAndAvailable` method does will probably change a little bit, but that previous commit will likely stay almost the same.

Comment (tahirali-csc) [reply pkg/controller/statefulset/stateful_set_control.go:762]: I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Comment (leomichalski) [reply pkg/controller/statefulset/stateful_set_control.go:762]: > I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Not expected. This flag (`--authentication-kubeconfig`) is set by the command specified here `spec.containers.command` in the "/etc/kubernetes/manifests/kube-controller-manager.yaml" file that I used. No "unknown flag" error at all.

After finding out the ID of the docker container running Kind, I run `docker cp kube-controller-manager.yaml SUBSTITUTE_WITH_THE_KIND_DOCKER_CONTAINER_ID:/etc/kubernetes/manifests/kube-controller-manager.yaml`. Then, I delete the old kube controller manager pod so it is recreated with the new manifest.

Comment (leomichalski) [reply pkg/controller/statefulset/stateful_set_control.go:762]: Here's the `kube-controller-manager.yaml` I used for reference:

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-controller-manager
    tier: control-plane
  name: kube-controller-manager
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-controller-manager
    - --allocate-node-cidrs=true
    - --authentication-kubeconfig=/etc/kubernetes/controller-manager.conf
    - --authorization-kubeconfig=/etc/kubernetes/controller-manager.conf
    - --bind-address=127.0.0.1
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --cluster-cidr=10.244.0.0/16
    - --cluster-name=lab
    - --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt
    - --cluster-signing-key-file=/etc/kubernetes/pki/ca.key
    - --controllers=*,bootstrapsigner,tokencleaner
    - --enable-hostpath-provisioner=true
    - --feature-gates=MaxUnavailableStatefulSet=true
    - --kubeconfig=/etc/kubernetes/controller-manager.conf
    - --leader-elect=true
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --root-ca-file=/etc/kubernetes/pki/ca.crt
    - --service-account-private-key-file=/etc/kubernetes/pki/sa.key
    - --service-cluster-ip-range=10.96.0.0/16
    - --use-service-account-credentials=true
    # image: registry.k8s.io/kube-controller-manager:v1.27.3
    image: registry.k8s.io/kube-controller-manager-amd64:v1.30.0-beta.0.13_bfce32798dbaef-dirty
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10257
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    name: kube-controller-manager
    resources:
      requests:
        cpu: 200m
    startupProbe:
      failureThreshold: 24
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10257
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    volumeMounts:
    - mountPath: /etc/ssl/certs
      name: ca-certs
      readOnly: true
    - mountPath: /etc/ca-certificates
      name: etc-ca-certificates
      readOnly: true
    - mountPath: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
      name: flexvolume-dir
    - mountPath: /etc/kubernetes/pki
      name: k8s-certs
      readOnly: true
    - mountPath: /etc/kubernetes/controller-manager.conf
      name: kubeconfig
      readOnly: true
    - mountPath: /usr/local/share/ca-certificates
      name: usr-local-share-ca-certificates
      readOnly: true
    - mountPath: /usr/share/ca-certificates
      name: usr-share-ca-certificates
      readOnly: true
  hostNetwork: true
  priority: 2000001000
  priorityClassName: system-node-critical
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  volumes:
  - hostPath:
      path: /etc/ssl/certs
      type: DirectoryOrCreate
    name: ca-certs
  - hostPath:
      path: /etc/ca-certificates
      type: DirectoryOrCreate
    name: etc-ca-certificates
  - hostPath:
      path: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
      type: DirectoryOrCreate
    name: flexvolume-dir
  - hostPath:
      path: /etc/kubernetes/pki
      type: DirectoryOrCreate
    name: k8s-certs
  - hostPath:
      path: /etc/kubernetes/controller-manager.conf
      type: FileOrCreate
    name: kubeconfig
  - hostPath:
      path: /usr/local/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-local-share-ca-certificates
  - hostPath:
      path: /usr/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-share-ca-certificates
status: {}
```

Comment (leomichalski) [reply pkg/controller/statefulset/stateful_set_control.go:762]: > I followed the above steps to load the tar file in the kind cluster but getting `unknown flag: --authentication-kubeconfig`. Is it expected?

Which command returns that error?

**Comment by leomichalski** [comment]
/hold

It's better to fix the StatefulSetMinReadySeconds beta feature before working on the StatefulSetMaxUnavailable alpha feature. Also because, as discussed here https://github.com/kubernetes/enhancements/pull/4474#discussion_r1478474701 , StatefulSetMinReadySeconds bugs are blocking StatefulSetMaxUnavailable graduation to beta.

**Comment by leomichalski** [comment]
I opened another PR just to fix the StatefulSetMinReadySeconds beta feature: https://github.com/kubernetes/kubernetes/pull/123975

**Pull request closed without merging**